lib/TH/generic/THTensorMath.c

```diff
@@ -1754,22 +1754,61 @@ void THTensor_(topk)(THTensor *rt_, THLongTensor *ri_, THTensor *t, long k, int
   long sliceSize = THTensor_(size)(t, dim);
   THArgCheck(k > 0 && k <= sliceSize, 2, "k not in range for dimension");
 
-  /* Just implement in terms of sort and narrow for now */
-  THTensor* tmpResults = THTensor_(new)();
-  THLongTensor* tmpIndices = THLongTensor_new();
+  THTensor *tmpResults = THTensor_(new)();
+  THTensor_(resize1d)(tmpResults, sliceSize);
+  real *tmp__data = THTensor_(data)(tmpResults);
 
-  THLongStorage* topKSize = THTensor_(newSizeOf)(t);
+  THLongTensor *tmpIndices = THLongTensor_new();
+  THLongTensor_resize1d(tmpIndices, sliceSize);
+  long *tmpi__data = THLongTensor_data(tmpIndices);
+
+  THLongStorage *topKSize = THTensor_(newSizeOf)(t);
   THLongStorage_set(topKSize, dim, k);
   THTensor_(resize)(rt_, topKSize, NULL);
   THLongTensor_resize(ri_, topKSize, NULL);
   THLongStorage_free(topKSize);
 
-  THTensor_(sort)(tmpResults, tmpIndices, t, dim, dir);
-  THTensor_(narrow)(tmpResults, NULL, dim, 0, k);
-  THLongTensor_narrow(tmpIndices, NULL, dim, 0, k);
+  if (dir) {
```

Comment:
Wouldn't it be simpler to do it with only one macro? Something like this should work:

```C
long K = 0;
TH_TENSOR_DIM_APPLY3(real, t, real, rt_, long, ri_, dim,
                     long i;
                     for(i = 0; i < sliceSize; i++)
                     {
                       tmp__data[i] = t_data[i*t_stride];
                       tmpi__data[i] = i;
                     }

                     if (dir) {
                       /* k largest elements, descending order (optional: see sorted) */
                       K = sliceSize - k;
                       if (K > 0)
                         THTensor_(quickselect)(tmp__data, tmpi__data, K - 1, sliceSize, 1);
                       if (sorted)
                         THTensor_(quicksortdescend)(tmp__data + K, tmpi__data + K, k, 1);
                     } else {
                       /* k smallest elements, ascending order (optional: see sorted) */
                       THTensor_(quickselect)(tmp__data, tmpi__data, k - 1, sliceSize, 1);
                       if (sorted)
                         THTensor_(quicksortascend)(tmp__data, tmpi__data, k - 1, 1);
                     }

                     for(i = 0; i < k; i++)
                     {
                       rt__data[i*rt__stride] = tmp__data[i + K];
                       ri__data[i*ri__stride] = tmpi__data[i + K];
                     })

```

Also is the copy of `sliceSize` necessary, because `quickselect` only works on contiguous Tensors?